ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java

```diff
@@ -0,0 +1,277 @@
+                    double clusterPosition[] = baseCluster.getPosition();
+                    double clusterEnergy = baseCluster.getEnergy();
+
+                    // To correct the energy, need the correct position - from previous iteration -
+                    // and the raw energy
+                    baseCluster.setPosition(clusterPosition);
+                    baseCluster.setEnergy(clusterEnergy_NC);
+                    ClusterEnergyCorrection2019.setCorrectedEnergy(ecal, baseCluster, isMC, addNoise);
+                    double clusterEnergy_C = baseCluster.getEnergy();
+
+                    // To correct the position, need the correct energy - from previous iteration -
+                    // and the raw position
+                    baseCluster.setEnergy(clusterEnergy);
+                    baseCluster.setPosition(clusterPosition_NC);
+                    ClusterPositionCorrection2019.setCorrectedPosition(baseCluster);
+
+                    // now set back the energy after correction (the position is already set)
+                    baseCluster.setEnergy(clusterEnergy_C);
+
+                    // absolute POS, relative in energy
+                    double deltaClusterEnergy = clusterEnergy - clusterEnergy_C;
+                    double deltaClusterX = clusterPosition[0] - baseCluster.getPosition()[0];
+                    double deltaClusterY = clusterPosition[1] - baseCluster.getPosition()[1];
+
+                    // A.C. : add a condition on the relative variation of the energy before-after
+                    // iteration
+                    // add a condition on the absolute variation of the position before-after
+                    // iteration
+                    // if all are satisfied, break the loop
+                    if ((Math.abs(deltaClusterEnergy / clusterEnergy) < deltaClusterEnergy_thr2019)
+                            && (Math.abs(deltaClusterX) < deltaClusterX_thr2019)
+                            && (Math.abs(deltaClusterY) < deltaClusterY_thr2019)) {
+                        break;
+                    }
+
+                }
+
+            } else {
+                // Apply PID based position correction, which should happen before final energy
+                // correction.
+                ClusterPositionCorrection.setCorrectedPosition(baseCluster);
+                ClusterEnergyCorrection.setCorrectedEnergy(ecal, baseCluster, isMC);
+            }
+        }
+    }
+
+    /**
+     * Apply HPS-specific energy and position corrections to a cluster with track
+     * information. In this case, ypos is the y position of the cluster, determined
+     * from clustering
+     * 
+     * @param cluster The input cluster.
+     */
+    public static void applyCorrections(double beamEnergy, HPSEcal3 ecal, Cluster cluster, double ypos, boolean isMC) {
+        if (cluster instanceof BaseCluster) {
+            BaseCluster baseCluster = (BaseCluster) cluster;
+
+            if (beamEnergy > 4.0) {
```

Comment:
The 2021 data set will most likely be taken at a lower beam energy, so this should be OK.

Reply:
So we don't expect to take any data beyond 2021?

Reply:
Yes, we do, but the 2021 data will be at energies less than 4.0 for which we already have corrections. Whether we ever run at energies much above 4.5 is still to be determined. We can come back to this if and when we need to, but I don't think this should hold up the current release.

Reply:
For any new run we will take, we will need to repeat the full set of: calibrations, position corrections, energy corrections. 

The reason is that these depend on the hit threshold, 30 MeV (2019) is very different than 5 MeV (2015 and 2016). Therefore, no matter what we have here now, we will need to repeat the full exercise, understand the detector behavior, validate corrections obtained from MC with real data, ecc... ecc... 

Only after that exercise will be completed, we will commit the new code to hps-java master branch. At that point, we will decide what do to with this "if beamEnergy > 4.0".  We can refactor all the code, or we can simply put another "if" based on the beamEnergy, or we can use the runRange (but what for MC). Since what we have now will work for 2019, that we still need to cook and analyze, I suggest to just move forward.

Reply:
I agree, and would just like to see the code written in a way that is more forward looking to putting in new additions like this. What I am asking for here is pretty simple and honestly should not take very much time. It honestly could have been done in roughly the amount of time it took both of you to type those responses to my comment.